File: abc133/abc133_b.cs
// https://atcoder.jp/contests/abc133/submissions/8263950

using System;
using System.Linq;

namespace CSharp
{
    class Program
    {
        public static void Main()
        {
            var input = Console.ReadLine().Split(' ').Select(int.Parse).ToArray();
            var N = input[0];
            var D = input[1];

            var X = new int[N][];
            for (int i = 0; i < N; i++)
            {
                X[i] = Console.ReadLine().Split(' ').Select(int.Parse).ToArray();
            }

            var result = 0;
            for (int i = 0; i < N; i++)
            {
                for (int j = i + 1; j < N; j++)
                {
                    var dist = 0;
                    for (int k = 0; k < D; k++)
                    {
                        dist += (X[i][k] - X[j][k]) * (X[i][k] - X[j][k]);
                    }
                    var t = (int)Math.Ceiling(Math.Sqrt(dist));
                    if (t * t == dist)
                    {
                        result++;
                    }
                }
            }
            Console.WriteLine(result);
        }
    }
}

File: abc/abc133/abc133_b.fs
// https://atcoder.jp/contests/abc133/submissions/11971637

let [| N; D |] = stdin.ReadLine().Split() |> Array.map int

let X =
    [| for i in [ 1..N ] do
        yield (stdin.ReadLine().Split() |> Array.map int) |]

let rec solver s i j =
    if i = N then s
    elif j = N then solver s (i + 1) (i + 2)
    else
        let d = Array.map2 (fun x y -> (x - y) * (x - y)) X.[i] X.[j] |> Array.sum

        let t =
            d
            |> float
            |> sqrt
            |> ceil
            |> int
        if t * t = d then solver (s + 1) i (j + 1)
        else solver s i (j + 1)

solver 0 0 1 |> stdout.WriteLine
